packages/utils/src/Signers.ts

```diff
@@ -0,0 +1,425 @@
+ *
+ * @param signers An array of signer interfaces.
+ * @param selectors One or more selector callbacks, receiving a signer as input and returning `true` in case it meets selection criteria.
+ * @returns The first signer for which all selectors returned `true`, or `undefined` if none meet selection criteria.
+ */
+export async function selectSigner(
+  signers: readonly SignerInterface[],
+  ...selectors: readonly SignerSelector[]
+): Promise<SignerInterface | undefined> {
+  return signers.find((signer) => {
+    return selectors.every((selector) => selector(signer))
+  })
+}
+
+/**
+ * Select signers based on (key) ids.
+ *
+ * @param ids Allowed signer/key ids to filter for.
+ * @returns A selector identifying signers whose id property is in `ids`.
+ */
+function byId(ids: readonly string[]): SignerSelector {
+  return ({ id }) => ids.includes(id)
+}
+/**
+ * Select signers based on algorithm identifiers.
+ *
+ * @param algorithms Allowed algorithms to filter for.
+ * @returns A selector identifying signers whose algorithm property is in `algorithms`.
+ */
+function byAlgorithm(algorithms: readonly string[]): SignerSelector {
+  return (signer) =>
+    algorithms.some(
+      (algorithm) => algorithm.toLowerCase() === signer.algorithm.toLowerCase()
+    )
+}
+/**
+ * Select signers based on the association of key ids with a given DID.
+ *
+ * @param didDocument DidDocument of the DID, on which the signer id must be listed as a verification method.
+ * @param options Additional optional filter criteria.
+ * @param options.verificationRelationship If set, the signer id must be listed under this verification relationship on the DidDocument.
+ * @param options.controller If set, only verificationMethods with this controller are considered.
+ * @returns A selector identifying signers whose id is associated with the DidDocument.
+ */
+function byDid(
+  didDocument: DidDocument,
+  {
+    controller,
+    verificationRelationship,
+  }: { verificationRelationship?: string; controller?: string } = {}
+): SignerSelector {
+  let eligibleVMs = didDocument.verificationMethod
+  // TODO: not super happy about this throwing; can I attach a diagnostics property to the returned function instead that will inform why this will never select a signer?
+  if (!Array.isArray(eligibleVMs) || eligibleVMs.length === 0) {
+    throw new DidError(
+      `DID ${didDocument.id} not fit for signing: No verification methods are associated with the signer DID document. It may be that this DID has been deactivated.`
+    )
+  }
+  if (controller) {
+    eligibleVMs = eligibleVMs.filter(
+      ({ controller: ctr }) => controller === ctr
+    )
+  }
+  // helps deal with relative DID URLs as ids
+  function absoluteId(id: string): DidUrl {
+    if (id.startsWith(didDocument.id)) {
+      return id as DidUrl
+    }
+    if (id.startsWith('#')) {
+      return `${didDocument.id}${id as UriFragment}`
+    }
+    return `${didDocument.id}#${id}`
```

Comment:
I could see the plain fragment ending up on a did document, even though it's technically not valid (at least if you don't add a `@base` to the document context that includes a #). This would catch this case?

Reply:
In other words, how do you explain somebody that 
```json
{
  "id": "did:kilt:4pnfkRn5UurBJTW92d9TaVLR2CqJdY4z5HPjrEbpGyBykare",
  "verificationMethod": [
    {
      "id": "#mykey-1",
      "type": "Multikey",
      "publicKeyMultibase": "z1234567234567234567234567"
    }
```
is a valid DID document but 
```json
{
  "id": "did:kilt:4pnfkRn5UurBJTW92d9TaVLR2CqJdY4z5HPjrEbpGyBykare",
  "verificationMethod": [
    {
      "id": "mykey-1",
      "type": "Multikey",
      "publicKeyMultibase": "z1234567234567234567234567"
    }
```
isn't? While the latter indeed isn't technically valid, I'd rather handle this generously than explain.

Reply:
I see. So, once again, you are mainly referring to JS developers. I'll take that.